source/Ship.cpp

```diff
@@ -3498,20 +3508,56 @@ const vector<weak_ptr<Ship>> &Ship::GetEscorts() const
 // cues and try to stay with it when it lands or goes into hyperspace.
 void Ship::AddEscort(Ship &ship)
 {
+	RegisterEscort(ship);
 	escorts.push_back(ship.shared_from_this());
 }
 
 
 
 void Ship::RemoveEscort(const Ship &ship)
 {
+	// Reset this value, we will re-scan the list of escorts to set the
+	// escorts velocity based on remaining active escorts.
+	escortsVelocity = -1.;
+
 	auto it = escorts.begin();
-	for( ; it != escorts.end(); ++it)
-		if(it->lock().get() == &ship)
-		{
-			escorts.erase(it);
-			return;
+	while(it != escorts.end())
+	{
+		auto escort = it->lock().get();
+		if(escort){
+			if(escort == &ship){
+				it = escorts.erase(it);
+				continue;
+			}
+			else
+				// Re-scan existing escorts to cache data like the
+				// escorts velocity.
+				RegisterEscort(ship);
 		}
+		++it;
```

Comment:
This will double increment on erasure, which means we will increment the `end()` iterator on removing the last escort, or skip the next ship in line.

(Minor style nit on the `if(escort){` too.)

Reply:
Isn't the double increment avoided by the `continue` statement on line 3530?

Reply:
I'll fix the style issue.

Reply:
Ah, i missed that line somehow. Indeed, it is.

An alternate formulation that reduces some indentation would be

```c++
for(auto it = escorts.begin(); it != escorts.end(); )
{
  auto escort = it.lock().get();
  if(escort == &ship)
    it = escorts.erase(it);
  else
  {
    if(escort)
      RegisterEscort(ship);
    ++it;
  }
}
```